larpandoracontent/LArTrackShowerId/TrackShowerIdFeatureTool.cc

```diff
@@ -199,147 +194,567 @@ void LinearFitFeatureTool::CalculateVariablesSlidingLinearFit(const pandora::Clu
-    featureVector.push_back(mipEnergy * m_mipCorrectionPerHit);
+    featureVector.push_back(vertexDistance);
 }
 
 //------------------------------------------------------------------------------------------------------------------------------------------
 
-StatusCode MipEnergyFeatureTool::ReadSettings(const TiXmlHandle xmlHandle)
+StatusCode ThreeDVertexDistanceFeatureTool::ReadSettings(const TiXmlHandle xmlHandle)
 {
+
     PANDORA_RETURN_RESULT_IF_AND_IF(STATUS_CODE_SUCCESS, STATUS_CODE_NOT_FOUND, !=, XmlHelper::ReadValue(xmlHandle,
-        "MipCorrectionPerHit", m_mipCorrectionPerHit));
+        "SlidingLinearFitWindow", m_slidingLinearFitWindow));
 
     return STATUS_CODE_SUCCESS;
 }
 
 //------------------------------------------------------------------------------------------------------------------------------------------
 //------------------------------------------------------------------------------------------------------------------------------------------
 
-void VertexDistanceFeatureTool::Run(SupportVectorMachine::DoubleVector &featureVector, const Algorithm *const pAlgorithm,
-    const pandora::Cluster *const pCluster)
+ThreeDOpeningAngleFeatureTool::ThreeDOpeningAngleFeatureTool()
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void ThreeDOpeningAngleFeatureTool::Run(SupportVectorMachine::DoubleVector &featureVector, const Algorithm *const pAlgorithm,
+    const pandora::ParticleFlowObject *const pInputPfo)
+{
+    if (PandoraContentApi::GetSettings(*pAlgorithm)->ShouldDisplayAlgorithmInfo())
+        std::cout << "----> Running Algorithm Tool: " << this->GetInstanceName() << ", " << this->GetType() << std::endl;
+
+    // Need the 3D clusters and hits to calculate PCA components
+    ClusterList threeDClusterList;
+    LArPfoHelper::GetThreeDClusterList(pInputPfo, threeDClusterList);
+
+    CaloHitList threeDCaloHitList;
+    LArPfoHelper::GetCaloHits(pInputPfo, TPC_3D, threeDCaloHitList);
+
+    if (threeDCaloHitList.empty())
+        return;
+
+    CartesianPointVector pointVectorStart, pointVectorEnd;
+
+    this->Divide3DCaloHitList(pAlgorithm, threeDCaloHitList, pointVectorStart, pointVectorEnd);
+    // Run the PCA analysis twice
+    CartesianVector centroidStart(0.f, 0.f, 0.f), centroidEnd(0.f, 0.f, 0.f);
+    LArPcaHelper::EigenVectors eigenVecsStart, eigenVecsEnd;
+    LArPcaHelper::EigenValues eigenValuesStart(0.f, 0.f, 0.f), eigenValuesEnd(0.f, 0.f, 0.f);
+
+    LArPcaHelper::RunPca(pointVectorStart, centroidStart, eigenValuesStart, eigenVecsStart);
+    LArPcaHelper::RunPca(pointVectorEnd, centroidEnd, eigenValuesEnd, eigenVecsEnd);
+
+    float openingAngle(this->OpeningAngle(eigenVecsStart.at(0), eigenVecsStart.at(1), eigenValuesStart));
+    float closingAngle(this->OpeningAngle(eigenVecsEnd.at(0), eigenVecsEnd.at(1), eigenValuesEnd));
+
+    featureVector.push_back(std::abs(openingAngle-closingAngle));
```

Comment:
We use `std::fabs(...)` by default, although I think the old abs vs. fabs issues are no more.